spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.sources.v2.DataSourceOptions;
+import org.apache.spark.sql.sources.v2.reader.streaming.MicroBatchReader;
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StreamingReader extends Reader implements MicroBatchReader {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReader.class);
+
+  private Offset startOffset;
+  private Offset endOffset;
+
+  private final Table table;
+  private final int maxTasksPerTrigger;
+
+  static class SnapshotTasks {
+    private final long snapshotId;
+    private final List<CombinedScanTask> tasks;
+    private final int size;
+
+    @VisibleForTesting
+    SnapshotTasks(long snapshotId, List<CombinedScanTask> tasks) {
+      this.snapshotId = snapshotId;
+      this.tasks = tasks;
+      this.size = tasks.size();
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    List<CombinedScanTask> tasks() {
+      return tasks;
+    }
+
+    int size() {
+      return size;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxTasksPerTrigger =
+      options.get("max-tasks-per-trigger").map(Integer::parseInt).orElse(1000);
+    Preconditions.checkArgument(this.maxTasksPerTrigger > 0,
+        "Option max-tasks-per-trigger should > 0");
+  }
+
+  @Override
+  public void setOffsetRange(Optional<Offset> start, Optional<Offset> end) {
+    startOffset = start.orElse(calculateInitialOffset());
+    endOffset = Optional.ofNullable(startOffset)
+        .map(s -> end.orElse(calculateEndOffset((StreamingOffset) s)))
+        .orElse(null);
+  }
+
+  @Override
+  public Offset getStartOffset() {
+    if (startOffset == null) {
+      throw new IllegalStateException("Start offset is not set");
+    }
+
+    return startOffset;
+  }
+
+  @Override
+  public Offset getEndOffset() {
+    if (endOffset == null) {
+      throw new IllegalStateException("End offset is not set");
+    }
+
+    return endOffset;
+  }
+
+  @Override
+  public Offset deserializeOffset(String json) {
+    return StreamingOffset.fromJson(json);
+  }
+
+  @Override
+  public void commit(Offset end) {}
+
+  @Override
+  public void stop() {}
+
+  @Override
+  protected List<CombinedScanTask> tasks() {
+    Preconditions.checkState(startOffset != null,
+        "Start offset is null when planning tasks, this probably because there's no snapshot in the table currently");
+    Preconditions.checkState(endOffset != null, "End offset is null when planning tasks");
```

Comment:
I don't think it needs to be handled in this commit, but it would be nice to return 0 tasks until there is data available in the table if it is new, instead of failing.

Reply:
Fixed.